processor.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	processor := &Processor{
+		client:         client,
+		subscriptionID: subscriptionID,
+		ctx:            ctx,
+		cancel:         cancel,
+		newStream: func(client *Client, id string, options *StreamOptions) streamAPI {
+			return NewStream(client, id, options)
+		},
+		timePerBatchPerStream: time.Duration(timePerBatchPerStream)}
+
+	for i := uint(0); i < options.StreamCount; i++ {
+		streamOptions := StreamOptions{
+			BatchLimit:           options.BatchLimit,
+			InitialRetryInterval: options.InitialRetryInterval,
+			MaxRetryInterval:     options.MaxRetryInterval,
+			CommitMaxElapsedTime: options.CommitMaxElapsedTime,
+			NotifyErr:            func(err error, duration time.Duration) { options.NotifyErr(i, err, duration) },
+			NotifyOK:             func() { options.NotifyOK(i) }}
+
+		processor.streamOptions = append(processor.streamOptions, streamOptions)
+	}
+
+	return processor
+}
+
+// A Processor for nakadi events. The Processor is a high level API for consuming events from
+// Nakadi. It can process event batches from multiple partitions (streams) and can be configured
+// with a certain event rate, that limits the number of processed events per minute. The cursors of
+// event batches that were successfully processed are automatically committed.
+type Processor struct {
+	sync.Mutex
+	client                *Client
+	subscriptionID        string
+	streamOptions         []StreamOptions
+	newStream             func(*Client, string, *StreamOptions) streamAPI
+	timePerBatchPerStream time.Duration
+	streams               []streamAPI
+	ctx                   context.Context
+	cancel                context.CancelFunc
+}
+
+// Start begins event processing. All event batches received from the underlying streams are passed to
+// the operation function. If the operation function terminates without error the respective cursor will
+// be automatically committed to Nakadi. If the operations terminates with an error, the underlying stream
+// will be halted and a new stream will continue to pass event batches to the operation function.
+//
+// Event processing will go on indefinitely unless the processor is stopped via its Stop method. Star will
+// return an error if the processor is already running.
+func (p *Processor) Start(operation func(int, []byte) error) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if len(p.streams) > 0 {
+		return errors.New("processor was already started")
+	}
+
+	p.streams = make([]streamAPI, len(p.streamOptions))
+
+	for i, o := range p.streamOptions {
+
+		go func(streamNo int, options StreamOptions) {
+			p.streams[streamNo] = p.newStream(p.client, p.subscriptionID, &options)
+
+			if p.timePerBatchPerStream > 0 {
+				initialWait := rand.Int63n(int64(p.timePerBatchPerStream))
+				select {
+				case <-p.ctx.Done():
+					return
+				case <-time.After(time.Duration(initialWait)):
+					// nothing
+				}
+			}
+
+			for {
```

Comment:
Having two loops in a function with all the breaks and continues disturbs reading it. It would be great to extract one loop to a separate function